controllers/alert_controller.go

```diff
@@ -21,146 +21,184 @@ import (
-		alert.Status.ObservedGeneration = alert.Generation
-		if err := r.patchStatus(ctx, req, alert.Status); err != nil {
-			return ctrl.Result{Requeue: true}, err
+	defer func() {
+		conditions.SetSummary(alert, meta.ReadyCondition)
+
+		patchOpts := []patch.Option{
+			patch.WithOwnedConditions{
+				Conditions: []string{
+					meta.ReadyCondition,
+					meta.ReconcilingCondition,
+					meta.StalledCondition,
+				},
+			},
 		}
-		log.Info("Alert initialised")
-	}
 
-	r.recordReadiness(ctx, alert)
+		if retErr == nil && (result.IsZero() || !result.Requeue) {
+			conditions.Delete(alert, meta.ReconcilingCondition)
 
-	return ctrl.Result{}, nil
-}
+			patchOpts = append(patchOpts, patch.WithStatusObservedGeneration{})
 
-func (r *AlertReconciler) SetupWithManager(mgr ctrl.Manager) error {
-	return ctrl.NewControllerManagedBy(mgr).
-		For(&v1beta1.Alert{}).
-		WithEventFilter(predicate.Or(predicate.GenerationChangedPredicate{}, predicates.ReconcileRequestedPredicate{})).
-		Complete(r)
+			readyCondition := conditions.Get(alert, meta.ReadyCondition)
+			switch readyCondition.Status {
+			case metav1.ConditionFalse:
+				// As we are no longer reconciling and the end-state is not ready, the reconciliation has stalled
+				conditions.MarkStalled(alert, readyCondition.Reason, readyCondition.Message)
+			case metav1.ConditionTrue:
+				// As we are no longer reconciling and the end-state is ready, the reconciliation is no longer stalled
+				conditions.Delete(alert, meta.StalledCondition)
+			}
+		}
+
+		if err := patchHelper.Patch(ctx, alert, patchOpts...); err != nil {
+			retErr = kerrors.NewAggregate([]error{retErr, err})
+		}
+
+		r.Metrics.RecordReadiness(ctx, alert)
+		r.Metrics.RecordDuration(ctx, alert, start)
+	}()
+
+	return r.reconcile(ctx, alert)
 }
 
-func (r *AlertReconciler) validate(ctx context.Context, alert v1beta1.Alert) error {
-	var provider v1beta1.Provider
-	providerName := types.NamespacedName{Namespace: alert.Namespace, Name: alert.Spec.ProviderRef.Name}
-	if err := r.Get(ctx, providerName, &provider); err != nil {
-		return fmt.Errorf("failed to get provider %s, error: %w", providerName.String(), err)
-	}
+func (r *AlertReconciler) reconcile(ctx context.Context, alert *v1beta1.Alert) (ctrl.Result, error) {
+	// Mark the resource as under reconciliation
+	conditions.MarkReconciling(alert, meta.ProgressingReason, "")
 
-	if !apimeta.IsStatusConditionTrue(provider.Status.Conditions, meta.ReadyCondition) {
-		return fmt.Errorf("provider %s is not ready", providerName.String())
+	// validate alert spec and provider
+	if err := r.validate(ctx, alert); err != nil {
+		conditions.MarkFalse(alert, meta.ReadyCondition, meta.FailedReason, err.Error())
+		return ctrl.Result{}, err
 	}
 
-	return nil
+	conditions.MarkTrue(alert, meta.ReadyCondition, meta.SucceededReason, v1beta1.InitializedReason)
+	ctrl.LoggerFrom(ctx).Info("Alert initialised")
```

Comment:
```suggestion
	ctrl.LoggerFrom(ctx).Info("Alert initialized")
```

to match the reason.